pm: device_runtime: Fix possible deadlock

ceolin opened the pull request:
- Fix deadlock when the device does not  implement pm
- Properly protect critical section using the spinlock instead of atomics.

Commit:
pm: device_runtime: Fix enable / disable runtime pm

Device pm runtime should only be enabled if device supports pm.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device_runtime: Fix atomic usage

Protect critical sessions using the spinlock available. The atomic
usage was not properly protecting the critical section and was
possible to have a race condition between the usage check and state
set.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

gmarull commented on include/device.h:
```diff
@@ -613,6 +613,29 @@ static inline bool device_is_ready(const struct device *dev)
+		return 0;
+	}
+#endif
+	return -ENOTSUP;
```
I'm not convinced about this API:

1. We're adding a PM related API to `device.h`
2. It should return a boolean, so it can be simplified by just `return dev->pm_control != NULL`. Which makes me think if it's really needed.

ceolin replied:
The information is needed. The function itself is other history, this is a helper that we may change in the future if the internal pm implementation changes. Right now, the device_runtime can just check if a device implements pm_control. Having a function seems cleaner. This can be moved to include/pm/device.h indeed.

gmarull commented on subsys/pm/device_runtime.c:
```diff
@@ -81,28 +85,27 @@ static void pm_work_handler(struct k_work *work)
 	 * is nothing we can do with this information. Just ignoring it.
 	 */
 	(void)k_condvar_broadcast(&dev->pm->condvar);
+	k_spin_unlock(&dev->pm->lock, key);
```
I think a single unlock point makes code more safe and readable, ie

```c
    ...
    goto end;

fsm_out:
     (void)k_condvar_broadcast(&dev->pm->condvar);

end:
     k_spin_unlock(&dev->pm->lock, key);
```

gmarull commented on subsys/pm/device_runtime.c:
```diff
@@ -127,13 +130,34 @@ static int pm_device_request(const struct device *dev,
+		goto end;
 	}
 
+	k_spin_unlock(&dev->pm->lock, key);
```
same comment here

ceolin commented:
Yep, that is wrong. I noticed just after sent it Friday. Have already fixed these issues, but I want to confirm that this is correct so I'm spending some time to test it properly. Thanks for reviewing it.

Commit:
pm: device_runtime: Add helper to wait on async ops

Add a function that properly uses a mutex to check a condition before
wait on the conditional variable.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
tests: pm: Add device_runtime test

- Test two threads changing states concurrently
- Test multiple calls to get/put
- Test async / sync API

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device_runtime: Fix sync path

Properly wait on conditional variable using the new API.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
samples: device_pm: Fix condvar usage

Use pm_device_wait() to properly wait an operation to finish.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

erwango commented on include/pm/device_runtime.h:
```diff
@@ -109,13 +109,31 @@ int pm_device_put(const struct device *dev);
 int pm_device_put_sync(const struct device *dev);
+
+/**
+ * @brief Wait on a device to finish an operation.
```
Can we clarify what kind of operation ?

ceolin replied:
Yep, will document it better. Operation here refers to put/get.

erwango commented on subsys/pm/device_runtime.c:
```diff
@@ -127,29 +124,48 @@ static int pm_device_request(const struct device *dev,
-	(void)k_condvar_wait(&dev->pm->condvar, &request_mutex, K_FOREVER);
-	k_mutex_unlock(&request_mutex);
+	k_spin_unlock(&dev->pm->lock, key);
+	pm_device_wait(dev, K_FOREVER);
```
Is that correct understanding that we'll wait forever here?
Or is there a specific behavior with K_FOREVER (in which case, it would need to be clarified in function header)

ceolin replied:
It will wait until the state transition finish. If there is no pending transition and the device is not in the middle of one it will return without wait.

ceolin commented on subsys/pm/device_runtime.c:
```diff
@@ -127,18 +125,36 @@ static int pm_device_request(const struct device *dev,
+	key = k_spin_lock(&dev->pm->lock);
+
+	if (!dev->pm->enable) {
+		ret = -ENOSYS;
```
it makes sense.

Commit:
pm: device_runtime: get rid of the spinlock

Protect critical sections using the mutex.
The mutex is required to use the conditional variable and since we
need to atomically check the pm state and the workqueue before wait
the condition, it is necessary to protect them using the same mutex.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device_runtime: Document isr and pre-kernel functions

Add notes for APIs that are allowed to be called from ISRs and
pre-kernel.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
samples: device_pm: Fix harness

Changing from spinlock to mutex makes the running thread be preempted
and the workqueue ran. This changed the prints order in this sample.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

nashif commented on tests/subsys/pm/device_runtime/src/main.c:
```diff
@@ -0,0 +1,240 @@
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	ret = pm_device_get(dev);
```
why is the thread calling this directly? Shouldn't you call driver's dummy_open()?

ceolin replied:
It could be, but then I would need to call pm_device_get_sync() to test the synchronous path ... To avoid further mistakes, I've just added open_sync and close_sync to the driver and calling them instead of using directly the get/put.

nashif replied:
well, create another driver function that uses get_sync to test this path, I think this will be confusing and error prone if we start calling those functions outside of a driver and we will have no way of testing the real use case.

nashif commented on tests/subsys/pm/device_runtime/src/main.c:
```diff
@@ -0,0 +1,240 @@
+	k_sem_give(&sem);
+
+	/* Block waiting for device operation conclude */
+	ret = pm_device_wait(dev, K_FOREVER);
```
same here, shouldn't this happen in the driver? The thread need not deal with this, this all needs to be done in the driver.

ceolin replied:
The purpose was testing the api, this is not a sample. Although I get the point, just created wrappers on driver for these functions instead of call them directly.

gmarull commented on include/pm/device_runtime.h:
```diff
@@ -32,6 +32,8 @@ extern "C" {
  * The device might be asynchronously suspended if runtime PM is enabled
  * when the device is not use.
  *
+ * @note This function is *pre-kernel-ok*.
```
We introduced a doc mechanism to list function properties:
```
@funcprops \isr_ok
```

ceolin replied:
ok, started using it then I saw some places using exactly this string and I couldn't find something like \pre_kernel_ok. Is it valid ?

ceolin replied:
just found the other alias.

gmarull commented on subsys/pm/device_runtime.c:
```diff
@@ -229,44 +225,42 @@ void pm_device_enable(const struct device *dev)
-	k_mutex_lock(&dev->pm->condvar_lock, K_FOREVER);
+	k_mutex_lock(&dev->pm->lock, K_FOREVER);
 	while ((k_work_delayable_is_pending(&dev->pm->work)) ||
 		(atomic_get(&dev->pm->state) == PM_DEVICE_STATE_SUSPENDING) ||
```
we should now get rid of the atomics, `mutex` should be the responsible to protect PM data

ceolin replied:
here is fine not using atomic, but there is one case where it is necessary. In `pm_device_request` in the sync path we need release the mutex before calling `pm_device_wait` and after this function return there is not mutex protecting the state.

But I can get rid of the atomic and use the mutex after the wait just to check the state.

gmarull commented on tests/subsys/pm/device_runtime/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+FILE(GLOB app_sources src/*.c)
+target_sources(app PRIVATE ${app_sources})
+
+target_include_directories(app PRIVATE
+	${ZEPHYR_BASE}/kernel/include
+	${ZEPHYR_BASE}/arch/${ARCH}/include)
```
are we using any of these private includes here?

ceolin replied:
this and the next comment is because I just copied the driver from another test.

ceolin replied:
leftover from another test. Just removing it.

gmarull commented on tests/subsys/pm/device_runtime/prj.conf:
```diff
@@ -0,0 +1,5 @@
+CONFIG_PM=y
+CONFIG_PM_DEVICE=y
+CONFIG_PM_DEVICE_RUNTIME=y
+CONFIG_MP_NUM_CPUS=1
```
why are we enforcing this condition?

ceolin replied:
To avoid SMP.

gmarull commented on tests/subsys/pm/device_runtime/src/main.c:
```diff
@@ -0,0 +1,240 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <sys/printk.h>
+#include <zephyr.h>
+#include <zephyr/types.h>
+#include <ztest.h>
+#include <ksched.h>
+#include <kernel.h>
+#include <pm/pm.h>
```
some headers look redundant here

Commit:
pm: device_runtime: Get rid of atomic for state

Since we are using mutex to protect critical sections and mutexes are
reentrant, it is possible to get rid of atomic for the state because
we can lock the mutex in device_pm_callback.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

gmarull commented on include/pm/device_runtime.h:
```diff
@@ -55,6 +59,8 @@ void pm_device_disable(const struct device *dev);
  * This API will asynchronously bring the device to resume state
  * if it not already in active state.
  *
+ * @funcprops \isr_ok \pre_kernel_ok
```
nit: 
```suggestion
 * @funcprops \isr_ok, \pre_kernel_ok
```

gmarull commented on include/pm/device_runtime.h:
```diff
@@ -86,6 +92,8 @@ int pm_device_get_sync(const struct device *dev);
  * This API asynchronously put the device to suspend state if
  * it not already in suspended state.
  *
+ * @funcprops \isr_ok \pre_kernel_ok
```
```suggestion
 * @funcprops \isr_ok, \pre_kernel_ok
```

nashif reviewed (changes requested):
more comments

ceolin commented on tests/subsys/pm/device_runtime/src/main.c:
```diff
@@ -0,0 +1,240 @@
+	k_sem_give(&sem);
+
+	/* Block waiting for device operation conclude */
+	ret = api->wait(dev);
```
then both open_sync() and and open() will have the same behavior .... Again, that is just to test device_runtime api.
If I put wait inside the open() then I'll not be able to test from the application multiple async calls and wait after many requests.

ceolin replied:
And btw, having two drivers one using sync api and the other using the async will screw the test because I won't be able to  mix sync/async calls in the same device.

nashif replied:
wait should not be part of open(), it will have to be called in the path of pm_device_get() and in device_runtime.c or where-ever the logic is implemented in the pm subsystem, if you have to call it outside of the driver APIs then that indicates  something is wrong with the workflow.

ceolin replied:
it is ... it is used in the sync path. But we expose this API, imagine that a device did an async request and later needs to check if this request finished or want to wait until it finishes ? That's the purpose of this function.

ceolin replied:
https://github.com/ceolin/zephyr/blob/pm/deadlock/subsys/pm/device_runtime.c#L155

ceolin replied:
We need to allow a driver knows when an operation has finished or wait to something to finish in the ASYNC case. This API is for this.
This is used internally in the synchronous path as you can see in the link.

nashif replied:
my point here is that the thread does not need to know about this and should not be calling wait(), sync vs async and pm handling all happens in the driver, not in the calling thread.

ceolin replied:
Yes, but this is not an application, is a test for that API .... Really, I'm not getting the point of over complicate the test.

ceolin replied:
I'm for instance doing multiple calls that are just to stress the API and not a real use case because this is a test to test that API.

nashif replied:
ok, that test is the only place we are using all of this, we are also missing documentation for this. What I want to see is how all of this would work outside of a test, so, you could add a new test that illustrates how this would work from an application, this new wait() API is just not clicking for me right now and need to see how it would work without it, in the application side of things.

ceolin replied:
Anas, look this https://github.com/ceolin/zephyr/blob/pm/deadlock/subsys/pm/device_runtime.c#L162
This helper is not only used in this test but also in `pm_device_get_sync` and `pm_device_put_sync`.

The API is documented in https://github.com/ceolin/zephyr/blob/pm/deadlock/include/pm/device_runtime.h#L124

I don't know exactly what is missing, but is useful if a device calls `pm_device_get` and then do other tasks it needs and then want to block until the request (`pm_device_get`) concludes.

ceolin commented on tests/subsys/pm/device_runtime/src/dummy_driver.c:
```diff
@@ -0,0 +1,101 @@
+
+static uint32_t device_power_state;
+
+static int dummy_wait(const struct device *dev)
```
because with this I can do

- open(async)
- open(async)
- close(async)
- wait()

If I put it in open/close I will end up having:
- open(async)
- - wait()
- open(async)
- - wait()
- close(async)
-- wait()

And the test will be different.

nashif commented on subsys/pm/device_runtime.c:
```diff
@@ -127,29 +122,48 @@ static int pm_device_request(const struct device *dev,
+		}
+	}
+
 	(void)k_work_schedule(&dev->pm->work, K_NO_WAIT);
```
Coming back to this, synchronous calls should not go through the work queue, this was a major flaw in the current implementation that we wanted to change.

ceolin replied:
The synchronous case can just wait the callback indeed. Let me just confirm.

nashif commented on tests/subsys/pm/device_runtime/src/main.c:
```diff
@@ -0,0 +1,237 @@
+			threadA_func, NULL, NULL, NULL,
+			K_PRIO_PREEMPT(1), 0, K_FOREVER);
+
+	/* Lets make threadB has higher priority than the workqueue
```
why do we need to worry about this? This seems like a potential for problems if we have to worry about system workq priority, we should test threads at different, but random priorities and not be concerned about workq

ceolin replied:
Otherwise we don't test concurrence, the workqueue will always run before the others threads .. Just because of this, if I don't do this we are not testing this corner case.

Commit:
pm: device_runtime: Don't use k_work on synchronous calls

Do not use worqueue on put/get synchronous calls.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
Close due https://github.com/zephyrproject-rtos/zephyr/pull/35646
